README.md
# bedrock-jwt-mongodb

CHANGELOG.md
# bedrock-jwt-mongodb ChangeLog

## 1.1.1 - 2018-03-26

### Changed
- Update `bedrock-mongodb` peer dependency.
- Update test dependencies.

## 1.1.0 - 2017-06-27

### Changed
- Upgrade `bedrock-key` peer dependency from 3.x to 4.x.

## 1.0.0 - 2017-01-20

### Added
- Initial release.
- See git history for changes.


test/test.js
/*!
 * Copyright (c) 2017 Digital Bazaar, Inc. All rights reserved.
 */
const bedrock = require('bedrock');
require('bedrock-jwt-mongodb');

require('bedrock-test');
bedrock.start();


test/test.config.js
/*
 * Copyright (c) 2017 Digital Bazaar, Inc. All rights reserved.
 */
const config = require('bedrock').config;
const path = require('path');

config.mocha.tests.push(path.join(__dirname, 'mocha'));

// MongoDB
config.mongodb.name = 'bedrock_jwt_mongodb_test';
config.mongodb.host = 'localhost';
config.mongodb.port = 27017;
config.mongodb.local.collection = 'bedrock_jwt_mongodb_test';
config.mongodb.dropCollections.onInit = true;
config.mongodb.dropCollections.collections = [];


test/mocha/10-api.js
var brJwt = require('bedrock-jwt-mongodb');

describe('API', () => {
  it('should work', done => {
    brJwt.provision({
      identifier: 'alpha', keyExpirationInSecs: 30
    }, (err, result) => {
      console.log('RRRRRRRRRr', result);
      done();
    });
  });
});


lib/store.js
/*!
 * Copyright (c) 2017 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';

const async = require('async');
const bedrock = require('bedrock');
const database = require('bedrock-mongodb');
const jwt = require('jsonwebtoken');
const BedrockError = bedrock.util.BedrockError;
const NamespaceHandlerWebKey = require('./namespace-handler-web-key');
const NamespaceHandlerHmac = require('./namespace-handler-hmac');

const logger = bedrock.loggers.get('app');

class Store {
  constructor() {
    const handlerWebKey = new NamespaceHandlerWebKey();
    this.handlers = {
      HS: new NamespaceHandlerHmac(),
      RS: handlerWebKey
      // TODO: Add ECDSA support
      // ES: handlerWebKey
    };
  }

  getNamespace(id, callback) {
    database.collections.jwtKeyStore.findOne(
      {id: database.hash(id)}, {namespace: true}, (err, record) => {
        if(err) {
          logger.error('Mongo error when trying to find key state.', err);
          return callback(new BedrockError(
            'Failed to find key state due to internal error.',
            'InternalError',
            {namespace: id, httpStatusCode: 500, 'public': true}, err));
        }
        if(!record) {
          return callback(new BedrockError(
            'Key state not found.',
            'NotFound',
            {namespace: id, httpStatusCode: 404, 'public': true}));
        }
        callback(null, record.namespace);
      });
  }

  /**
   * Initialize a namespaced keystore.
   *
   * @param namespace the namespace to initialize.
   * @param options the options to use:
   *   id the ID for the namespace.
   *   algorithm the JWT signing algorithm.
   *   clockToleranceInSecs the acceptable clock skew.
   *   tokenTtlInSecs the JWT token TTL.
   *   [key] the key identifier or a key object with:
   *     id the key ID.
   *
   * @param callback(err) called once the operation completes.
   */
  provision(options, callback) {
    const self = this;

    // build namespace from options
    const namespace = {
      id: options.id,
      algorithm: options.algorithm,
      clockToleranceInSecs: options.clockToleranceInSecs,
      tokenTtlInSecs: options.tokenTtlInSecs,
      state: null
    };
    if('key' in options) {
      if(typeof options.key === 'string') {
        namespace.key = options.key;
      } else if(typeof options.key === 'object') {
        namespace.key = options.key.id;
      } else {
        return callback(
          new TypeError('`options.key` must be a string or object.'));
      }
    }

    // get namespace handler
    let handler;
    try {
      handler = self._getNamespaceHandler(namespace.algorithm);
    } catch(e) {
      return callback(new BedrockError(
        'Unsupported algorithm.', 'NotFound',
        {public: false, namespace: options}));
    }

    async.auto({
      createState: handler.createState.bind(handler, {namespace: namespace}),
      storeNamespace: ['createState', (callback, results) => {
        namespace.state = results.createState;
        const now = Date.now();
        const record = {
          id: database.hash(namespace.id),
          meta: {
            created: now,
            updated: now
          },
          namespace: namespace
        };
        database.collections.jwtKeyStore.insert(
          record, database.writeOptions, (err, result) => {
            if(err) {
              if(database.isDuplicateError(err)) {
                // namespace is already provisioned
                // TODO: ensure record in database matches given options
                return callback();
              }
              logger.error(
                '[jwt-mongodb] Mongo error when trying to insert namespace.',
                err);
              return callback(new BedrockError(
                'Failed to insert namespace due to internal error.',
                'InternalError',
                {namespace: options.id, httpStatusCode: 500, 'public': true},
                err));
            }
            callback(null, result.ops[0]);
          });
      }]
    }, callback);
  }

  /**
   * Create a JWT.
   *
   * @param options the options to use:
   *   namespace the namespace to use for the signing operation.
   *   payload the payload to be included in the JWT.
   * @param callback(err, token) called once the operation completes.
   */
  sign(options, callback) {
    const self = this;
    async.auto({
      getNamespace: self.getNamespace.bind(self, options.namespace),
      sign: ['getNamespace', (callback, results) => {
        let handler;
        try {
          handler = self._getNamespaceHandler(results.getNamespace.algorithm);
        } catch(e) {
          return callback(e);
        }
        handler.sign({
          namespace: results.getNamespace,
          payload: options.payload
        }, callback);
      }]
    }, (err, results) => callback(err, results ? results.sign : null));
  }

  /**
   * Verifies a JWT.
   *
   * @param token the token to be verified.
   * @param callback(err, tokenPayload) called once the operation completes.
   */
  verify(token, callback) {
    let header;
    let decodedToken;
    try {
      decodedToken = jwt.decode(token, {complete: true});
      header = decodedToken.header;
    } catch(err) {
      return callback(err);
    }

    if(!header.alg.startsWith('HS')) {
      return callback(null, new Error('Unsupported algorithm.'));
    }

    const self = this;
    const parsedKid = self._parse(header.kid);

    async.auto({
      getNamespace: callback =>
        self.getNamespace(parsedKid.namespace, callback),
      verify: ['getNamespace', (callback, results) => {
        let handler;
        try {
          handler = self._getNamespaceHandler(results.getNamespace.algorithm);
        } catch(e) {
          return callback(e);
        }
        handler.verify(
          token, {
            namespace: results.getNamespace,
            keyId: parsedKid.keyId
          }, callback);
      }]
    }, (err, results) => callback(err, results ? results.verify : null));
  }

  _parse(kid) {
    const parts = kid.split(':');
    return {
      namespace: parts[0],
      keyId: parts[1]
    };
  }

  _getNamespaceHandler(algorithm) {
    const handler = this.handlers[algorithm.substr(0, 2)];
    if(!handler) {
      throw new Error('Unsupported algorithm.');
    }
    return handler;
  }
}

module.exports = Store;


lib/namespace-handler.js
/*!
 * Copyright (c) 2017 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';

const _ = require('lodash');
const async = require('async');
const bedrock = require('bedrock');
const jwt = require('jsonwebtoken');

class NamespaceHandler {
  /**
   * Creates any custom state information for the given namespace during
   * provisioning.
   *
   * @param options the options to use:
   *   namespace the namespace:
   *     id the identifier for the namespace.
   *     algorithm the JWT signing algorithm.
   *     tokenTtlInSecs the JWT token TTL.
   *     state any custom namespace state.
   *     [key] an optional key identifier for this namespace.
   * @param callback(err, state) called once the operation completes.
   */
  createState(options, callback) {
    callback(new Error(
      'NamespaceHandler#createState must be overriden by subclass.'));
  }

  /**
   * Gets a key from the given options (or elsewhere).
   *
   * @param options the options to use:
   *   namespace the namespace:
   *     id the identifier for the namespace.
   *     algorithm the JWT signing algorithm.
   *     tokenTtlInSecs the JWT token TTL.
   *     state any custom namespace state.
   *     [key] an optional key identifier for this namespace.
   * @param callback(err, {id: keyId, material: keyMaterial}) called
   *   once the operation completes.
   */
  getKey(options, callback) {
    callback(new Error(
      'NamespaceHandler#getKey must be overriden by subclass.'));
  }

  /**
   * Verifies a JWT.
   *
   * @param token the token to be verified.
   * @param options the options to use:
   *   namespace the namespace to verify with.
   *   keyId the namespace-specific key ID to use, parsed from the token's
   *     `kid` value (any namespace identifier in the `kid` has been removed).
   * @param callback(err, tokenPayload) called once the operation completes.
   */
  verify(token, options, callback) {
    callback(new Error(
      'NamespaceHandler#verify must be overriden by subclass.'));
  }

  /**
   * Create a JWT.
   *
   * @param options the options to use:
   *   namespace the namespace:
   *     id the identifier for the namespace.
   *     algorithm the JWT signing algorithm.
   *     tokenTtlInSecs the JWT token TTL.
   *     state any custom namespace state.
   *     [key] an optional key identifier for this namespace.
   *   payload the payload to sign and include in the JWT.
   *
   * @param callback(err, token) called once the operation completes.
   */
  sign(options, callback) {
    const self = this;
    const payload = bedrock.util.clone(options.payload);
    async.auto({
      getKey: self.getKey.bind(self, {namespace: options.namespace}),
      create: ['getKey', (callback, results) => {
        const key = results.getKey;
        const nowInSecs = Math.floor(Date.now() / 1000);
        const notAfter = (nowInSecs + options.namespace.tokenTtlInSecs);
        _.assign(payload, {exp: notAfter, iat: nowInSecs});
        const token = jwt.sign(payload, key.material, {
          algorithm: options.namespace.algorithm,
          header: {
            kid: options.namespace.algorithm.startsWith('HS') ?
              (options.namespace.id + ':' + key.id) : key.id
          }
        });
        return callback(null, token);
      }]
    }, (err, results) => callback(err, results ? results.create : null));
  }
}

module.exports = NamespaceHandler;


lib/namespace-handler-hmac.js
/*!
 * Copyright (c) 2017 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';

const _ = require('lodash');
const async = require('async');
const bedrock = require('bedrock');
const crypto = require('crypto');
const database = require('bedrock-mongodb');
const jwt = require('jsonwebtoken');
const BedrockError = bedrock.util.BedrockError;
const NamespaceHandler = require('./namespace-handler');

const logger = bedrock.loggers.get('app');

class NamespaceHandlerHmac extends NamespaceHandler {
  /**
   * Creates any custom state information for the given namespace during
   * provisioning.
   *
   * @param options the options to use:
   *   namespace the namespace:
   *     id the identifier for the namespace.
   *     algorithm the JWT signing algorithm.
   *     tokenTtlInSecs the JWT token TTL.
   *     state any custom namespace state.
   *     [key] an optional key identifier for this namespace.
   * @param callback(err, state) called once the operation completes.
   */
  createState(options, callback) {
    const nowInSecs = Math.floor(Date.now() / 1000);
    const state = {
      previousKey: null,
      key: {
        id: '' + nowInSecs,
        data: this._generateHmacKey(),
        created: nowInSecs,
        expires: nowInSecs + options.namespace.tokenTtlInSecs +
          options.namespace.clockToleranceInSecs
      }
    };
    callback(null, state);
  }

  /**
   * Gets a key from the given options (or elsewhere).
   *
   * @param options the options to use:
   *   namespace the namespace:
   *     id the identifier for the namespace.
   *     algorithm the JWT signing algorithm.
   *     tokenTtlInSecs the JWT token TTL.
   *     state any custom namespace state.
   *     [key] an optional key identifier for this namespace.
   * @param callback(err, {id: keyId, material: keyMaterial}) called
   *   once the operation completes.
   */
  getKey(options, done) {
    const self = this;

    async.auto({
      getNamespace: callback => {
        if(typeof options.namespace !== 'string') {
          return callback(null, options.namespace);
        }
        self._get(options.namespace, callback);
      },
      rotate: ['getNamespace', (callback, results) => {
        const state = results.getNamespace.state;
        // NOTE: requires clock sync amongst nodes
        const nowInSecs = Math.floor(Date.now() / 1000);
        if(state.key.expires > nowInSecs) {
          // key not expired, nothing to rotate
          return callback(null, state);
        }
        // key expired, generate a new one
        state.previousKey = state.key;
        state.key = {
          id: '' + nowInSecs,
          data: self._generateHmacKey(),
          created: nowInSecs,
          expires: nowInSecs + state.tokenTtlInSecs +
            state.tokenClockToleranceInSecs
        };
        logger.verbose('[jwt-mongodb] recycling HMAC key...');
        self._update(options.namespace.id, state, {
          query: {
            'namespace.state.key.id': state.previousKey.id
          }
        }, (err, updated) => {
          if(err) {
            logger.error(
              'Mongo error when trying to get key state.', err);
            return callback(err);
          }
          if(!updated) {
            // another process updated; clear namespace, loop, and try again
            logger.verbose(
              '[jwt-mongodb] another process recycled HMAC key.');
            options = _.assign({}, options, {namespace: options.namespace.id});
            return process.nextTick(self.getKey.bind(self, options, done));
          }
          callback(null, state);
        });
      }],
      getKey: ['rotate', (callback, results) => {
        callback(null, {
          id: results.rotate.key.id,
          material: new Buffer(results.rotate.key.data, 'base64')
        });
      }]
    }, (err, results) => done(err, results ? results.getKey : null));
  }

  /**
   * Verifies a JWT.
   *
   * @param token the token to be verified.
   * @param options the options to use:
   *   namespace the namespace to verify with.
   *   keyId the namespace-specific key ID to use, parsed from the token's
   *     `kid` value (any namespace identifier in the `kid` has been removed).
   * @param callback(err, tokenPayload) called once the operation completes.
   */
  verify(token, options, callback) {
    // get key that matches key ID
    const state = options.namespace.state;
    let key;
    if(options.keyId === state.key.id) {
      key = state.key;
    } else if(state.previousKey && options.keyId === state.previousKey.id) {
      key = state.previousKey;
    } else {
      return callback(new Error('Invalid key identifier in token.'));
    }
    // decode and verify the token
    jwt.verify(
      token, new Buffer(key.data, 'base64'), {
        clockTolerance: options.namespace.clockToleranceInSecs
      }, callback);
  }

  _get(id, callback) {
    database.collections.jwtKeyStore.findOne(
      {id: database.hash(id)}, {namespace: true}, (err, record) => {
        if(err) {
          logger.error('Mongo error when trying to find key state.', err);
          return callback(new BedrockError(
            'Failed to find key state due to internal error.',
            'InternalError',
            {key: id, httpStatusCode: 500, 'public': true}, err));
        }
        if(!record) {
          return callback(new BedrockError(
            'Key not found.',
            'NotFound',
            {key: id, httpStatusCode: 404, 'public': true}));
        }
        callback(null, record.namespace);
      });
  }

  _update(id, state, options, callback) {
    if(typeof options === 'function') {
      callback = options;
      options = {};
    }
    options = options || {};
    const query = _.assign({}, options.query || {});
    query.id = database.hash(id);
    database.collections.jwtKeyStore.update(
      query, {
        $set: {
          'meta.updated': Date.now(),
          'namespace.state': state
        }
      }, database.writeOptions, (err, result) => {
        if(err) {
          logger.error('Mongo error when trying to update key state.', err);
          return callback(new BedrockError(
            'Failed to update key state due to internal error.',
            'InternalError',
            {namespace: id, httpStatusCode: 500, 'public': true},
            err));
        }
        callback(null, result.result.n > 0);
      });
  }

  _generateHmacKey() {
    return crypto.randomBytes(16).toString('base64');
  }
} // end class

module.exports = NamespaceHandlerHmac;


lib/namespace-handler-web-key.js
/*!
 * Copyright (c) 2017 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';

const bedrock = require('bedrock');
const brKey = require('bedrock-key');
const BedrockError = bedrock.util.BedrockError;
const NamespaceHandler = require('./namespace-handler');

class NamespaceHandlerWebKey extends NamespaceHandler {
  /**
   * Creates any custom state information for the given namespace during
   * provisioning.
   *
   * @param options the options to use:
   *   namespace the namespace:
   *     id the identifier for the namespace.
   *     algorithm the JWT signing algorithm.
   *     tokenTtlInSecs the JWT token TTL.
   *     state any custom namespace state.
   *     [key] an optional key identifier for this namespace.
   * @param callback(err, state) called once the operation completes.
   */
  createState(options, callback) {
    // all we need to do here is make sure the key is valid
    this.getKey(options, err => callback(err, {}));
  }

  /**
   * Gets a key from the given options (or elsewhere).
   *
   * @param options the options to use:
   *   namespace the namespace:
   *     id the identifier for the namespace.
   *     algorithm the JWT signing algorithm.
   *     tokenTtlInSecs the JWT token TTL.
   *     state any custom namespace state.
   *     [key] an optional key identifier for this namespace.
   * @param callback(err, {id: keyId, material: keyMaterial}) called
   *   once the operation completes.
   */
  getKey(options, callback) {
    const key = options.namespace.key;
    brKey.getPublicKey({id: key}, null, (err, publicKey, meta, privateKey) => {
      if(err) {
        return callback(new BedrockError('Invalid signing key specified.',
          'InvalidKey', {key: key}, err));
      }
      // ensure that the key has not been revoked
      if(publicKey.sysStatus !== 'active') {
        return callback(new BedrockError(
          'The specified signing key has been revoked.',
          'InvalidKey', {key: key}));
      }
      callback(null, {id: key, material: privateKey.privateKeyPem});
    });
  }
} // end class

module.exports = NamespaceHandlerWebKey;


lib/index.js
/*
 * Copyright (c) 2017 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';
const _ = require('lodash');
const async = require('async');
const bedrock = require('bedrock');
const crypto = require('crypto');
const jwt = require('jsonwebtoken');
const database = require('bedrock-mongodb');
const BedrockError = bedrock.util.BedrockError;

let _keyState = null;

// load config defaults
require('./config');

const api = {};
module.exports = api;

const logger = bedrock.loggers.get('app');

// TODO: make responding to this event optional
bedrock.events.on(
  'bedrock-authn-did-jwt.config.keyStore', (strategy, callback) => {
    strategy.setStore(api, callback);
  });

bedrock.events.on('bedrock-mongodb.ready', callback => {
  async.waterfall([
    callback => {
      database.openCollections(['jwtKeyStore'], callback);
    },
    callback => {
      database.createIndexes([{
        collection: 'jwtKeyStore',
        fields: {id: 1},
        options: {unique: true, background: false}
      }], callback);
    }], callback);
});

/**
 * Initialize a namespaced keystore.
 *
 * @param options the options to use:
 *   algorithm the JWT signing algorithm to use.
 *   clockToleranceInSecs the clockTolerance to use in during JWT verification.
 *   namespace the namespace to use.
 *   provision true for provisioning operation.
 *   ttlInSecs the JWT ttl.
 * @param callback(err, state) called once the operation completes.
 */
api.provision = function(options, callback) {
  options = _.assign({}, options, {provision: true});
  _getState(options, err => callback(err));
};

/**
 * Create a JWT.
 *
 * @param options the options to use:
 *   namespace the namespace for the keystore.
 *   payload the payload to sign and include in the JWT.
 *
 * @param callback(err, state) called once the operation completes.
 */
api.sign = function(options, callback) {
  async.auto({
    getState: callback => {
      return _getState({namespace: options.namespace}, callback);
    },
    create: ['getState', (callback, results) => {
      const state = results.getState;
      const nowInSecs = Math.floor(Date.now() / 1000);
      const notAfter = (nowInSecs + state.tokenTtlInSecs);
      const payload = _.assign(
        {}, options.payload, {exp: notAfter, iat: nowInSecs});
      const token = jwt.sign(payload, state.key.data, {
        algorithm: state.algorithm,
        header: {kid: state.key.id}
      });

      callback(null, token);
    }]
  }, (err, results) => {
    callback(err, results ? results.create : null);
  });
};

/**
 * Verify a JWT.
 *
 * @param options the options to use:
 *   namespace the namespace for the keystore.
 *   token the token to be verified.
 *
 * @param callback(err, state) called once the operation completes.
 */
api.verify = function(options, callback) {
  async.auto({
    get: callback => {
      _getState({namespace: options.namespace}, callback);
    },
    verify: ['get', (callback, results) => {
      const state = results.get;
      let token;
      try {
        // parse header
        const header = jwt.decode(options.token, {complete: true}).header;

        // get key that matches key ID
        let key;
        if(header.kid === state.key.id) {
          key = state.key;
        } else if(state.previousKey && header.kid === state.previousKey.id) {
          key = state.previousKey;
        } else {
          throw new Error('Invalid key identifier in token.');
        }
        // decode and verify the token
        token = jwt.verify(
          options.token, key.data, {
            algorithms: [state.key.algorithm],
            clockTolerance: state.clockToleranceInSecs
          });
      } catch(e) {
        return callback(e, false);
      }
      callback(null, token);
    }]
  }, (err, results) => {
    callback(err, results ? results.verify : false);
  });
};

function _get(id, callback) {
  database.collections.jwtKeyStore.findOne(
    {id: database.hash(id)}, {state: true}, (err, record) => {
      if(err) {
        logger.error('Mongo error when trying to find key state.', err);
        return callback(new BedrockError(
          'Failed to find key state due to internal error.',
          'InternalError',
          {namespace: id, httpStatusCode: 500, 'public': true}, err));
      }
      if(!record) {
        return callback(new BedrockError(
          'Key state not found.',
          'NotFound',
          {namespace: id, httpStatusCode: 404, 'public': true}));
      }
      callback(null, record.state, record.meta);
    });
}

function _insert(id, state, callback) {
  const now = Date.now();
  const record = {
    id: database.hash(id),
    meta: {
      created: now,
      updated: now
    },
    state: state
  };
  database.collections.jwtKeyStore.insert(
    record, database.writeOptions, (err, result) => {
      if(err) {
        if(database.isDuplicateError(err)) {
          return callback(new BedrockError(
            'The key state is a duplicate and could not be added.',
            'DuplicateRecord', {
              namespace: id,
              httpStatusCode: 409,
              'public': true
            }));
        }
        logger.error('Mongo error when trying to insert key state.', err);
        return callback(new BedrockError(
          'Failed to insert key state due to internal error.',
          'InternalError',
          {namespace: id, httpStatusCode: 500, 'public': true},
          err));
      }
      callback(null, result.ops[0]);
    });
}

function _update(id, state, options, callback) {
  if(typeof options === 'function') {
    callback = options;
    options = {};
  }
  options = options || {};
  const query = _.assign({}, options.query || {});
  query.id = database.hash(id);
  database.collections.jwtKeyStore.update(
    query, {
      $set: {
        'meta.updated': Date.now(),
        'state': state
      }
    }, database.writeOptions, (err, result) => {
      if(err) {
        logger.error('Mongo error when trying to update key state.', err);
        return callback(new BedrockError(
          'Failed to update key state due to internal error.',
          'InternalError',
          {namespace: id, httpStatusCode: 500, 'public': true},
          err));
      }
      callback(null, result.result.n > 0);
    });
}

/**
 * Gets the cached key state information, retrieving it from the database
 * and recycling HMAC keys as needed.
 *
 * @param options the options to use:
 *   algorithm the JWT signing algorithm to use.
 *   clockToleranceInSecs the clockTolerance to use in during JWT verification.
 *   namespace the namespace to use.
 *   provision true for provisioning operation.
 *   ttlInSecs the JWT ttl.
 * @param callback(err, state) called once the operation completes.
 */
function _getState(options, callback) {
  const done = callback;
  async.auto({
    get: callback => {
      // use cached key state if available, expiration checked later
      if(_keyState) {
        return callback(null, _keyState);
      }
      // get key state from database
      _get(options.namespace, (err, state) => {
        if(err && err.name === 'NotFound' && options.provision) {
          return callback(null, null);
        }
        callback(err, state);
      });
    },
    ensureExists: ['get', (callback, results) => {
      if(results.get) {
        return callback(null, results.get);
      }
      // attempt to insert new state
      const nowInSecs = Math.floor(Date.now() / 1000);
      const state = {
        id: options.namespace,
        previousKey: null,
        tokenTtlInSecs: options.ttlInSecs,
        tokenClockToleranceInSecs: options.clockToleranceInSecs,
        key: {
          id: '' + nowInSecs,
          algorithm: options.algorithm,
          data: _generateHmacKey(),
          created: nowInSecs,
          expires: nowInSecs + options.ttlInSecs + options.clockToleranceInSecs
        }
      };
      _insert(options.namespace, state, err => {
        if(err && err.name === 'DuplicateRecord') {
          // another process inserted; clear cache, loop, and try again
          _keyState = null;
          return process.nextTick(() => {
            _getState(done);
          });
        }
        return callback(err, state);
      });
    }],
    update: ['ensureExists', (callback, results) => {
      const state = results.ensureExists;
      // NOTE: requires clock sync amongst nodes
      const nowInSecs = Math.floor(Date.now() / 1000);
      if(state.key.expires > nowInSecs) {
        // key not expired, nothing to do
        return callback(null, state);
      }
      // key expired, generate a new one
      state.previousKey = state.key;
      state.key = {
        id: '' + nowInSecs,
        algorithm: state.previousKey.algorithm,
        data: _generateHmacKey(),
        created: nowInSecs,
        expires: nowInSecs + state.tokeTtlInSecs +
          state.tokenClockToleranceInSecs
      };
      logger.verbose('[jwt-mongodb] recycling HMAC key...');
      _update(options.namespace, state, {
        query: {
          'state.key.id': state.previousKey.id
        }
      }, (err, updated) => {
        if(err) {
          logger.error(
            'Mongo error when trying to get key state.', err);
          return callback(err);
        }
        if(!updated) {
          // another process updated; clear cache, loop, and try again
          logger.verbose('[jwt-mongodb] another process recycled HMAC key.');
          _keyState = null;
          return process.nextTick(() => {
            _getState(done);
          });
        }
        callback(null, state);
      });
    }],
    decode: ['update', (callback, results) => {
      const state = results.update;
      state.key.data = new Buffer(state.key.data, 'base64');
      if(state.previousKey) {
        state.previousKey.data = new Buffer(
          state.previousKey.data, 'base64');
      }
      // cache key state
      _keyState = state;
      callback(null, state);
    }]
  }, err => {
    callback(err, _keyState);
  });
}

function _generateHmacKey() {
  return crypto.randomBytes(16).toString('base64');
}
